plenum/server/consensus/ordering_service.py

```diff
@@ -2282,48 +2285,81 @@ def process_view_change_started(self, msg: ViewChangeStarted):
 
     def _update_old_view_preprepares(self, pre_prepares: List[PrePrepare]):
         for pp in pre_prepares:
-            view_no = pp.originalViewNo if f.ORIGINAL_VIEW_NO.nm in pp else pp.viewNo
+            view_no = get_original_viewno(pp)
             self.old_view_preprepares[(view_no, pp.ppSeqNo, pp.digest)] = pp
 
     def process_new_view_checkpoints_applied(self, msg: NewViewCheckpointsApplied):
         result, reason = self._validate(msg)
         if result != PROCESS:
             return result, reason
 
-        if not self.is_master:
-            return DISCARD, "not master"
-
+        missing_batches = []
         for batch_id in msg.batches:
             pp = self.old_view_preprepares.get((batch_id.pp_view_no, batch_id.pp_seq_no, batch_id.pp_digest))
             if pp is None:
-                # TODO: implement correct re-sending logic
-                # self._request_pre_prepare(three_pc_key=(batch_id.view_no, batch_id.pp_seq_no))
-                continue
-            new_pp = updateNamedTuple(pp, viewNo=self.view_no, originalViewNo=batch_id.pp_view_no)
+                missing_batches.append(batch_id)
+            else:
+                self._process_pre_prepare_from_old_view(pp)
 
-            # PrePrepare is accepted from the current Primary only
-            sender = generateName(self._data.primary_name, self._data.inst_id)
-            self.process_preprepare(new_pp, sender)
+        if missing_batches:
+            self._request_old_view_pre_prepares(missing_batches)
 
         return PROCESS, None
 
+    def process_old_view_preprepare_request(self, msg: OldViewPrePrepareRequest, sender):
+        result, reason = self._validate(msg)
+        if result != PROCESS:
+            return result, reason
+
+        old_view_pps = []
+        for batch_id in msg.batch_ids:
+            batch_id = BatchID(*batch_id)
+            pp = self.old_view_preprepares.get((batch_id.pp_view_no, batch_id.pp_seq_no, batch_id.pp_digest))
+            if pp is not None:
+                old_view_pps.append(pp)
+        rep = OldViewPrePrepareReply(self._data.inst_id, old_view_pps)
+        self._send(rep, dst=[getNodeName(sender)])
+
+    def process_old_view_preprepare_reply(self, msg: OldViewPrePrepareReply, sender):
+        result, reason = self._validate(msg)
+        if result != PROCESS:
+            return result, reason
+
+        for pp_dict in msg.preprepares:
```

Comment:
Should we check the order of received preprepares? It's can help with reducing "missing Pre-Prepare" requests.

Reply:
Since we use lists (not dicts) in `OldViewPrePrepareRequest` and `OldViewPrePrepareReply`, the order of PPs will be preserved.
Please have a look at `test_send_reply_on_old_view_pre_prepares_request` and `test_process_preprepare_on_old_view_pre_prepares_reply` where we make sure that result is in expected order

Reply:
ok